Add Check Support for Java 17 Sealed Classes: InterfaceMemberImpliedModifier

child of https://github.com/checkstyle/checkstyle/issues/14969
I have read check documentation: https://checkstyle.org/checks/modifier/interfacememberimpliedmodifier.html#InterfaceMemberImpliedModifier

---

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="InterfaceMemberImpliedModifier">

        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
sealed interface I permits A {
    String s = null; // 3 violations
    void m(); // 2 violations
    sealed interface J permits A {  // 2 violations
        void m();  // 2 violations
    }
}
final class A implements I, I.J {
    @Override
    public void m() {
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:2:5: Implied modifier 'final' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:2:5: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:2:5: Implied modifier 'static' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:3:5: Implied modifier 'abstract' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:3:5: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:4:5: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:4:5: Implied modifier 'static' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:5:9: Implied modifier 'abstract' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] D:\CS\test\src\Test.java:5:9: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
Audit done.
Checkstyle ends with 9 errors.
PS D:\CS\test> 

```

---

the check works as intended in sealed interfaces.

We are good to add input here and move on, the compiler handles the logic to make sure that we are explicit within a sealed hierarchy:
```
➜  src cat Test.java 
public class Test {
    public sealed interface I permits A, B {
    }

    non-sealed class A implements I {}
    class B implements I {} // could be considered a violation if it compiled

    class C extends A{}
    class D extends B{}
}
➜  src javac Test.java 
Test.java:6: error: sealed, non-sealed or final modifiers expected
    class B implements I {} // could be considered a violation if it compiled
    ^
1 error



```

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/InterfaceMemberImpliedModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/InterfaceMemberImpliedModifierCheckTest.java
@@ -385,4 +385,25 @@ public class InterfaceMemberImpliedModifierCheckTest
         }
     }
 
+    @Test
+    public void testSealedClassInInterface() throws Exception {
+        final String[] expected = {
+            "18:5: " + getCheckMessage(MSG_KEY, "final"),
+            "18:5: " + getCheckMessage(MSG_KEY, "public"),
+            "18:5: " + getCheckMessage(MSG_KEY, "static"),
+            "23:5: " + getCheckMessage(MSG_KEY, "abstract"),
+            "23:5: " + getCheckMessage(MSG_KEY, "public"),
+            "27:5: " + getCheckMessage(MSG_KEY, "public"),
+            "27:5: " + getCheckMessage(MSG_KEY, "static"),
+            "31:9: " + getCheckMessage(MSG_KEY, "abstract"),
+            "31:9: " + getCheckMessage(MSG_KEY, "public"),
+            "36:5: " + getCheckMessage(MSG_KEY, "static"),
+            "36:5: " + getCheckMessage(MSG_KEY, "public"),
+            "40:5: " + getCheckMessage(MSG_KEY, "public"),
+            "40:5: " + getCheckMessage(MSG_KEY, "static"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputInterfaceMemberImpliedModifierSealedClasses.java"),
+            expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/interfacememberimpliedmodifier/InputInterfaceMemberImpliedModifierSealedClasses.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/interfacememberimpliedmodifier/InputInterfaceMemberImpliedModifierSealedClasses.java
@@ -1,0 +1,62 @@
+/*
+InterfaceMemberImpliedModifier
+violateImpliedPublicField = (default)true
+violateImpliedStaticField = (default)true
+violateImpliedFinalField = (default)true
+violateImpliedPublicMethod = (default)true
+violateImpliedAbstractMethod = (default)true
+violateImpliedPublicNested = (default)true
+violateImpliedStaticNested = (default)true
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.modifier.interfacememberimpliedmodifier;
+
+public interface InputInterfaceMemberImpliedModifierSealedClasses {
+    String s = null;
+    // 3 violations above:
+    //                  'Implied modifier 'final' should be explicit.'
+    //                  'Implied modifier 'public' should be explicit.'
+    //                  'Implied modifier 'static' should be explicit.'
+    void m();
+    // 2 violations above:
+    //                  'Implied modifier 'abstract' should be explicit.'
+    //                  'Implied modifier 'public' should be explicit.'
+    sealed interface J permits A {
+    // 2 violations above:
+    //                  'Implied modifier 'public' should be explicit.'
+    //                  'Implied modifier 'static' should be explicit.'
+        void m();
+        // 2 violations above:
+        //                  'Implied modifier 'abstract' should be explicit.'
+        //                  'Implied modifier 'public' should be explicit.'
+    }
+    sealed class B permits A, D { }
+    // 2 violations above:
+    //                  'Implied modifier 'public' should be explicit.'
+    //                  'Implied modifier 'static' should be explicit.'
+    non-sealed class D extends B { }
+    // 2 violations above:
+    //                  'Implied modifier 'public' should be explicit.'
+    //                  'Implied modifier 'static' should be explicit.'
+}
+
+interface InputInterfaceMemberImpliedModifierSealedClassesCorrected {
+    final public static String s = null;
+    abstract public void m();
+    public  static  sealed interface J permits A {
+        abstract public void m();
+    }
+    public static sealed class B permits D { }
+    public static non-sealed class D extends B { }
+}
+
+final class A extends InputInterfaceMemberImpliedModifierSealedClasses.B
+        implements InputInterfaceMemberImpliedModifierSealedClasses.J,
+        InputInterfaceMemberImpliedModifierSealedClassesCorrected.J {
+
+    @Override
+    public void m() {}
+}

